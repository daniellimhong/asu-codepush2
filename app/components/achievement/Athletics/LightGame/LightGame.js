import React from "react";
import {
  View,
  StyleSheet,
  Text,
  TouchableOpacity,
  ImageBackground,
  Image,
  AppState
} from "react-native";
import MaterialIcon from "react-native-vector-icons/MaterialIcons";
import Permissions from "react-native-permissions";
import {
  responsiveHeight,
  responsiveWidth,
  responsiveFontSize
} from "react-native-responsive-dimensions";
import PropTypes from "prop-types";
import { HeaderQuick } from "../../Header/HeaderQuick";

let Torch = null;

export class LightGame extends React.PureComponent {
  constructor(props) {
    super(props);
    this.isTorchOn = false;
    this.canRunGame = false;
    this._interval = null;
    this._timeout = null;
    this.state = {
      cameraPermission: "undetermined",
      toggle: false
    };
    Torch = require("react-native-torch").default;
  }

  componentDidMount() {
    this.checkCameraPermissions();
    AppState.addEventListener("change", this.handleAppStateChange);
  }
  componentWillUnmount() {
    this.disableGame();
    AppState.removeEventListener("change", this.handleAppStateChange);
  }

  handleAppStateChange = nextAppState => {
    if (nextAppState.match(/inactive|background/)) {
      this.disableGame();
    } else {
      this.startGame();
    }
  };

  render() {
    let { navigate, goBack } = this.props.navigation;
    let leftNav = (
      <View
        style={{
          flexDirection: "row",
          flex: 1,
          justifyContent: "flex-start",
          alignItems: "center"
        }}
      >
        <TouchableOpacity
          onPress={() => {
            this.disableGame();
            goBack();
          }}
        >
          <MaterialIcon
            name="navigate-before"
            size={40}
            color="white"
            accessibilityLabel="Back"
            accessibilityRole="button"
          />
        </TouchableOpacity>
      </View>
    );

    return (
      <View>
        <HeaderQuick
          left={leftNav}
          theme="dark"
          navigation={this.props.navigation}
          title={"Light It Up!"}
        />
        <LightGameScreen
          toggleLight={this.toggleGame}
          toggle={this.state.toggle}
        />
      </View>
    );
  }

  /**
   * Check permissions of game
   */
  checkCameraPermissions() {
    if (this.props.testing === true) {
      //Do nothing if testing
    } else {
      Permissions.checkMultiple(["camera"]).then(response => {
        this.setState({
          cameraPermission: response.camera
        });
      });
    }
  }

  /**
   * Toggle the lights game on or off
   */
  toggleGame = () => {
    if (
      this.state.cameraPermission == "authorized" ||
      this.state.cameraPermission == "undetermined"
    ) {
      if (!this.canRunGame) {
        this.startGame();
      } else {
        this.disableGame();
      }
    }
  };

  /**
   * Set the interval and timeout of the lights game
   */
  startGame() {
    this.canRunGame = true;
    this._interval = setInterval(() => {
      this.swapTorch();
    }, 300);

    /*this._timeout = setTimeout(() => {
      this.disableGame()
    },40000)*/
  }

  /**
   * Disable the lights game, turning off the light and clearing the interval and timeout
   */
  disableGame() {
    this.canRunGame = false;
    this.killInterval();
    if (this.isTorchOn) {
      this.swapTorch();
    }
  }

  /**
   * Toggle the light on/off
   */
  swapTorch() {
    this.isTorchOn = !this.isTorchOn;
    Torch.switchState(this.isTorchOn);
  }

  /**
   * Clear interval & timeout
   */
  killInterval() {
    clearInterval(this._interval);
    clearTimeout(this._timeout);
  }
}

class LightGameScreen extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      gold: true
    };
  }

  static defaultProps = {
    toggle: false,
    toggleLight: () => null
  };

  _timeout = null;

  _interval = null;
  componentDidMount() {
    this.toggle();
  }
  componentDidUpdate(prevProps, prevState) {
    if (prevProps.toggle !== this.props.toggle) {
      if (this.props.toggle) {
        this.toggle();
      } else {
        this.killInterval();
      }
    }
  }

  toggle = () => {
    if (this._interval == null) {
      this.props.toggleLight();
      this._interval = setInterval(() => {
        this.swapScreen();
      }, 600);

      // this._timeout = setTimeout(() => {
      //   this.killInterval()
      // },30000)
    } else {
      this.props.toggleLight();
      this.killInterval();
      this._timeout = null;
      this._interval = null;
    }
  };

  swapScreen() {
    this.setState({
      gold: !this.state.gold
    });
  }

  killInterval() {
    clearInterval(this._interval);
    clearTimeout(this._timeout);
  }

  render() {
    return (
      <TouchableOpacity
        onPress={() => {
          //this.toggle();
        }}
      >
        {this.state.gold ? <LightGameGold /> : <LightGameMaroon />}
      </TouchableOpacity>
    );
  }
}

class LightGameGold extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      background: require("./assets/background-gold.jpg"),
      goDevils: require("./assets/godevils1a2x.png"),
      pitchfork: require("./assets/Pitchfork1a2x.png")
    };
  }

  render() {
    return (
      <ImageBackground
        source={this.state.background}
        style={{ width: responsiveWidth(100), height: responsiveHeight(100) }}
        resizeMethod="resize"
      >
        <View
          style={{
            flex: 1,
            alignItems: "center",
            marginTop: responsiveHeight(5)
          }}
        >
          <Image
            style={{ height: responsiveHeight(27) }}
            source={this.state.goDevils}
            resizeMode="contain"
            resizeMethod="resize"
          />
          <Image
            style={{ height: responsiveHeight(50) }}
            source={this.state.pitchfork}
            resizeMode="contain"
            resizeMethod="resize"
          />
        </View>
      </ImageBackground>
    );
  }
}

class LightGameMaroon extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      background: require("./assets/background-maroon.jpg"),
      goDevils: require("./assets/godevilsa2x.png"),
      pitchfork: require("./assets/Pitchforka2x.png")
    };
  }

  render() {
    return (
      <ImageBackground
        source={this.state.background}
        style={{ width: responsiveWidth(100), height: responsiveHeight(100) }}
        resizeMethod="resize"
      >
        <View
          style={{
            flex: 1,
            alignItems: "center",
            marginTop: responsiveHeight(5)
          }}
        >
          <Image
            style={{ height: responsiveHeight(27) }}
            source={this.state.goDevils}
            resizeMode="contain"
            resizeMethod="resize"
          />
          <Image
            style={{ height: responsiveHeight(50) }}
            source={this.state.pitchfork}
            resizeMode="contain"
            resizeMethod="resize"
          />
        </View>
      </ImageBackground>
    );
  }
}
