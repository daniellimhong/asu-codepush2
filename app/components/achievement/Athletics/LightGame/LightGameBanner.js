import React from "react";
import { View, StyleSheet, Text, TouchableOpacity } from "react-native";
import FontAwesome from "react-native-vector-icons/FontAwesome";
import Permissions from "react-native-permissions";
import {
  responsiveHeight,
  responsiveWidth,
  responsiveFontSize
} from "react-native-responsive-dimensions";
import PropTypes from "prop-types";

let Torch = null;

export class LightGameBanner extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      gameDay: false
    };
    Torch = require("react-native-torch").default;
  }

  componentDidMount() {
    this.context
      .getAdminSettings()
      .then(settings => {
        if (settings && settings.gameDay == true) {
          this.setState({
            gameDay: settings.gameDay
          });
        }
      })
      .catch(e => {
        console.log("Error getting game day settings", e);
      });
  }

  render() {
    let { navigate } = this.props.navigation;
    // if(this.state.gameDay === true){
    return (
      <TouchableOpacity
        onPress={() => {
          navigate("LightGame");
        }}
      >
        <View style={styles.banner}>
          <View style={styles.bannerContent}>
            <Text
              style={styles.bannerText}
              accessibilityLabel="Light it up"
              accessibilityRole="button"
            >
              LIGHT IT UP!
            </Text>
            <FontAwesome
              name="angle-double-right"
              color="white"
              size={responsiveFontSize(5)}
              style={{
                marginLeft: responsiveWidth(2)
              }}
            />
          </View>
        </View>
      </TouchableOpacity>
    );
    // } else {
    //   return null
    // }
  }
}

const styles = StyleSheet.create({
  banner: {
    backgroundColor: "#941E41",
    alignItems: "center",
    justifyContent: "center",
    padding: responsiveWidth(10),
    margin: responsiveWidth(5)
  },
  bannerContent: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center"
  },
  bannerText: {
    color: "white",
    fontSize: responsiveFontSize(5.5),
    fontWeight: "bold",
    fontFamily: "Roboto"
  }
});

LightGameBanner.contextTypes = {
  getAdminSettings: PropTypes.func
};
