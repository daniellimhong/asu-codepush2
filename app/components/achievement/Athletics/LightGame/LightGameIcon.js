import React from "react";
import { View, StyleSheet, TouchableOpacity } from "react-native";
import FontAwesome from "react-native-vector-icons/FontAwesome";
import Permissions from "react-native-permissions";

let Torch = null;

export default class LightGameIcon extends React.PureComponent {
  constructor(props) {
    super(props);
    this.isTorchOn = false;
    this.canRunGame = false;
    this._interval = null;
    this._timeout = null;
    this.state = {
      cameraPermission: "undetermined"
    };
    Torch = require("react-native-torch").default;
  }

  static defaultProps = {
    testing: false
  };

  componentDidMount() {
    this.checkCameraPermissions();
  }

  render() {
    return (
      <View style={styles.iconBox}>
        <TouchableOpacity
          onPress={() => {
            this.toggleGame();
          }}
        >
          <FontAwesome name="flash" size={30} color={"white"} />
        </TouchableOpacity>
      </View>
    );
  }

  /**
   * Check permissions of game
   */
  checkCameraPermissions() {
    if (this.props.testing === true) {
      //Do nothing if testing
    } else {
      Permissions.checkMultiple(["camera"]).then(response => {
        this.setState({
          cameraPermission: response.camera
        });
      });
    }
  }

  /**
   * Toggle the lights game on or off
   */
  toggleGame() {
    if (
      this.state.cameraPermission == "authorized" ||
      this.state.cameraPermission == "undetermined"
    ) {
      if (!this.canRunGame) {
        this.startGame();
      } else {
        this.disableGame();
      }
    }
  }

  /**
   * Set the interval and timeout of the lights game
   */
  startGame() {
    this.canRunGame = true;
    this._interval = setInterval(() => {
      this.swapTorch();
    }, 300);

    this._timeout = setTimeout(() => {
      this.disableGame();
    }, 30000);
  }

  /**
   * Disable the lights game, turning off the light and clearing the interval and timeout
   */
  disableGame() {
    this.canRunGame = false;
    this.killInterval();
    if (this.isTorchOn) {
      this.swapTorch();
    }
  }

  /**
   * Toggle the light on/off
   */
  swapTorch() {
    this.isTorchOn = !this.isTorchOn;
    Torch.switchState(this.isTorchOn);
  }

  /**
   * Clear interval & timeout
   */
  killInterval() {
    clearInterval(this._interval);
    clearTimeout(this._timeout);
  }
}

const styles = StyleSheet.create({
  iconBox: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
    borderColor: "#242424",
    borderBottomWidth: 1
  },
  drawerItem: {
    padding: 20,
    width: "100%",
    borderBottomWidth: 1,
    borderColor: "#313131"
  },
  drawerText: {
    color: "white"
  },
  drawerBottom: {
    padding: 20,
    width: "100%"
  }
});
