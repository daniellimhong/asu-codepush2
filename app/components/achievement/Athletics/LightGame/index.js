export * from "./LightGameBanner";
export * from "./LightGameIcon";
export * from "./LightGame";
