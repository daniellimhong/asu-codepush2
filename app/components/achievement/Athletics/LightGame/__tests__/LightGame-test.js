import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import LightGame from '../LightGame';
import Enzyme from "enzyme";

it('renders correctly', () => {
  Enzyme.configure({ adapter: new Adapter() })
  const wrapper = shallow(
    <LightGame testing={true}/>
  );
  expect(wrapper).toMatchSnapshot();
});