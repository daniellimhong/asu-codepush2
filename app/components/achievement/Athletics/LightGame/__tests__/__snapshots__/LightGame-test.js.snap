// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "borderBottomWidth": 1,
      "borderColor": "#242424",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
  >
    <Icon
      allowFontScaling={false}
      color="white"
      name="flash"
      size={30}
    />
  </TouchableOpacity>
</View>
`;
